### PR TITLE
AddAllBaseTypes - check HasType before calling GetSchemaType to prevent exception on missing types

### PR DESCRIPTION
**Pull request by bzbetty**
was throwing an exception when using AddAllBaseTypes which didn't seem useful, now it just silently skips unknown base classes

### CODE REVIEW

**Comment by bzbetty** [comment]
Wondering if this feature should be opt in.

Either by a new argument to

AddMutationsFrom<class>(autoAddInputTypes: false, addMethodsWithoutMutationAttribute: false, argumentBinding: bindingtype.MutationArgument | bindingTyping.Direct)

Or via the mutation attribute (less preferable as I don't use it now)

[GraphQLMutation(bindingtype.MutationArgument | bindingTyping.Direct)]

Thoughts?

**Comment by bzbetty** [comment]
Also trying to think of what tests still need adding

* behaviour with autoAddInputTypes = false
* Behaviour when not in class with other mutations (eg one of the other mutations in the test class may have bound something for me)
* does it still work with method level injection/dbcontext

**Comment by bzbetty** [comment]
Also feel like @breyed may want to input on this one since it came from his original ticket.

File: src/EntityGraphQL/Schema/SchemaType.cs [hunk 1]
```diff
@@ -308,18 +308,18 @@ public SchemaType<TBaseType> RequiresAnyPolicy(params string[] policies)
 
 
         public override ISchemaType AddAllBaseTypes()
-        {
-            var baseType = Schema.GetSchemaType(TypeDotnet.BaseType, null);
-            if (baseType != null)
+        {            
+            if (Schema.HasType(TypeDotnet.BaseType))
             {
+                var baseType = Schema.GetSchemaType(TypeDotnet.BaseType, null);
```
[the remark this answers is not in the document]
Comment (bzbetty) [reply src/EntityGraphQL/Schema/SchemaType.cs:314]: Yes if they're already in the schema, currently it errors if they're not, which I guess might be nicer in some cases (silent failure can be bad).  However for me I often use marker interfaces or base classes to add in functionality like soft deletes or automatic change date tracking which I wouldn't want exposed.

Was actually wonderign if the automatic dbcontext mapping should do something with base types too (currently it marks abstract classes/interfaces as gql interfaces but doesn't set the base class).

[the remark this answers is not in the document]
Comment (lukemurray) [reply src/EntityGraphQL/Schema/SchemaType.cs:314]: Can you explain what you mean by "set a base class"?

Or to expand on it
The auto context mapper (FromObject) will now add base classes & interfaces as interfaces in the schema. But is will not mark the type as inheriting from/implementing those types?

"setting a base class" marks in the schema as that type inheriting from the base class?

[the remark this answers is not in the document]
Comment (bzbetty) [reply src/EntityGraphQL/Schema/SchemaType.cs:314]: ah sorry.

"set a base class" I mean in the GQL SDL they are output as implementing one or more interfaces.

[the remark this answers is not in the document]
Comment (lukemurray) [reply src/EntityGraphQL/Schema/SchemaType.cs:314]: OK. Just walking through this as in 3.0.0 I'm making breaking changes so if we want to make this different it is a good chance.

In the example added to the docs - read the comments I added

```c#
public abstract class Character {
    public int Id { get; set; }
    public string Name { get; set; }
    public IEnumerable<Character> Friends { get; set; }
    public IEnumerable<Episode> AppearsIn { get; set; }
}

public class Human : Droid {
  public IEnumerable<Starship> starships { get; set; }
  public int TotalCredits { get; set;}
}

public class Droid : Character {
    public stirng PrimaryFunction { get; set;}
}

// adds the interface Character
schema.AddInterface<Character>(name: "Character", description: "represents any character in the Star Wars trilogy") 
    // adds all fields/properties on Character to the schema type Character
    .AddAllFields(); 

// add schema type Human
schema.AddType<Human>("") 
    // add all fields/properties on Human to the schema type Human
    .AddAllFields() 
    // adds all base types as types to the schema. In this case Droid? Also add the fields of that base type? And marks Human as implementing Droid ?
    .AddAllBaseTypes(); 

// I think this is meant to be updating the type as Droid was added above?
schema.AddType<Droid>("") 
    // tells the schema that Droid implements Character
    .AddBaseType<Character>(); 
```

Are all those assumptions correct?

[the remark this answers is not in the document]
Comment (bzbetty) [reply src/EntityGraphQL/Schema/SchemaType.cs:314]: I didn't intend AddAllBaseTypes to add base types to the schema, I wouldn't mind it doing so but if not maybe it needs to be renamed (ImplementsInterface<Type>? AddBaseTypesAsInterfaces() )

Wondering if we should add all fields from the interface automatically to the child type.

So a couple of things I got wrong in the original implementation that we might need to fix up (seem to work with the tooling I use, but is apparently against the spec)

* Interfaces can't inherit/implement other interfaces  
* You can only implement interfaces not types
 
So in the example above I think Human implmenting droid would be illegal as droid already implements another type.

you can implement multiple interfaces using the & operator, which might be a way around the above problem (except that Droid can't be an interface)

`type Human implements Droid & Character`

* The implementation for .AddBaseType<Character>();  currently doesn't do any c# inheritence check, so it's probably possible to do illegal things, I didn't actually write a test to see what it does if the objects are completely different.


Wondering if we should have a thread for 3.0 changes?

[the remark this answers is not in the document]
Comment (lukemurray) [reply src/EntityGraphQL/Schema/SchemaType.cs:314]: Happy to start a general discussion about 3.0.0. I wish milestones has comments https://github.com/EntityGraphQL/EntityGraphQL/milestone/1

In general the idea was to clean up the obsolete APIs - i.e. make the breaking changes.

**Comment by lukemurray** [comment]
The spec does allow for interfaces to implement other interfaces. https://spec.graphql.org/October2021/#sec-Interfaces.Interfaces-Implementing-Interfaces

I'd like to follow the spec. Which means we should do

- [ ] You can only implement interfaces not types
- [ ] Allow multiple implements via the `&` operator on a type or interface
- [ ] Do not allow a type to implement another type

Methods/API wise how about (thinking out loud here, I think you've you'd this feature more so keen for feedback)

- Keep `schema.AddInterface<Character>()` to directly add interfaces to the schema
- Add `SchemaType.Implements<Character>(bool addIfNotInSchema = true, bool addAllFieldsIfAddedToSchema = true)` to allow you tell the type any interfaces it implements. I.e. `schema.AddInterface<Character>().Implements<T>()` or `schema.AddType<Human>().Implements<Character>()`.
- Rename `SchemaType.AddAllBaseTypes` 
  - `ImplementAllBaseTypes(bool autoAddAllFieldsInInterface = true)` - Adds the base classes & interfaces of the dotnet type to the schema if not there and makes the schema type implement them
- Remove `SchemaType.AddBaseType<>()` and use `.Implements()`

The above code example would be

```c#
public abstract class Character {
    public int Id { get; set; }
    public string Name { get; set; }
    public IEnumerable<Character> Friends { get; set; }
    public IEnumerable<Episode> AppearsIn { get; set; }
}

public class Human : Character {
  public IEnumerable<Starship> starships { get; set; }
  public int TotalCredits { get; set; }
}

public class Droid : Character {
    public stirng PrimaryFunction { get; set;}
}

// add the interface Character
schema.AddInterface<Character>(name: "Character", description: "represents any character in the Star Wars trilogy") 
    // adds all fields/properties on Character to the schema type Character
    .AddAllFields(); 

schema.AddType<Human>("") 
    .AddAllFields() 
    .Implements<Character>(); 

schema.AddType<Droid>("") 
    .Implements<Character>(); 
```

Or without adding the Interface directly you could

```c#
schema.AddType<Human>("") 
    .AddAllFields() 
    .ImplementAllBaseTypes(true, true); 

schema.AddType<Droid>("") 
    .ImplementAllBaseTypes(true, true);
```

Thoughts? Am I missing something? It is very similar but uses the Implement wording which matches graphic schema language?

**Comment by bzbetty** [comment]
my bad, thought i'd read an interface couldn't implement another somewhere.  Glad i'm wrong there.

Yes I like the naming you've suggested.

**Comment by lukemurray** [comment]
I don't think this PR is required any more. Lucky it was small :) 

I'm going to make the changes outlined here against #155

**Pull request closed without merging**